src/segment.hpp

```diff
@@ -87,79 +80,44 @@ class segment {
         return false;
     }
 
-    segment_properties<value_type> properties;
-
-    void add_mechanism(parameter_list p)
-    {
-        auto it = std::find_if(
-            mechanisms_.begin(), mechanisms_.end(),
-            [&p](parameter_list const& l){return l.name() == p.name();}
-        );
-        if(it!=mechanisms_.end()) {
-            throw std::out_of_range(
-                "attempt to add a mechanism parameter set to a segment which has an existing mechanism with the same name"
-            );
-        }
-
-        mechanisms_.push_back(std::move(p));
+    util::optional<mechanism_spec&> mechanism(const std::string& name) {
+        auto it = std::find_if(mechanisms_.begin(), mechanisms_.end(),
+            [&](mechanism_spec& m) { return m.name()==name; });
+        return it==mechanisms_.end()? util::nothing: util::just(*it);
     }
 
-    parameter_list& mechanism(std::string const& n)
-    {
-        auto it = std::find_if(
-            mechanisms_.begin(), mechanisms_.end(),
-            [&n](parameter_list const& l){return l.name() == n;}
-        );
-        if(it==mechanisms_.end()) {
-            throw std::out_of_range(
-                "attempt to access a parameter that is not defined in a segment"
-            );
+    void add_mechanism(mechanism_spec mech) {
```

Comment:
Has the expected behavior changed when adding a mechanism with the same name as a mechanism that is already present? i.e. from throwing an error to replacing the existing mechanism?

Reply:
Yes, it has.